### DISCUSSION TITLE
Looking forward

### DISCUSSION THREAD

**Discussion by codeart1st** [maplibre/maplibre-rs#18] [Show and tell]
I started to implement the same idea some weeks ago as a private project for me to learn rust and experiment a little bit with wgpu.

To focus on the renderer itself I decided to use OpenLayers as a base.

Rendering is completly inside an OffscreenCanvas with rust and wgpu-rs. Vector tile parsing with prost, triangulation with earcutr and multithreading with wasm-bindgen-rayon.

Looking forward to this project now. Nice to see that there is also some work going on in maplibre space for this future tech.

![164468814-9791e905-23a5-4054-bca7-e03901b27b91](https://user-images.githubusercontent.com/581407/164492523-8b1d0fc5-b41f-4f67-9c6f-0847f7a57a0b.png)

**Comment by maxammann** [comment]
I'm also looking forward on this project :) Sounds interesting, do you have your source code public?

**Reply by codeart1st** [reply]
Yeah I will move my project from my personal GitLab Instance to GitHub under MIT license this weekend. It's not something that should be used somewhere and not this polished as this experiment here, but I think it's better for new technology to be developed in the public space. Also to show browser vendors the demand of this kind of new features.

In my opinion rust, wasm and wgpu will be the common used stack for multiplatform graphics. And as a result, for mapping libraries as well.

**Reply by codeart1st** [reply]
https://github.com/codeart1st/wgpu-layers

As I said, nothing polished.

My goal is to build a renderer on top of openlayers ecosystem. And add/replace features later.

**Reply by maxammann** [reply]
Looked through the project. Looks very interesting!
I actually already had the idea to use an OffscreenCanvas for rendering text. Sadly the browser support is not good.
There is actually a similar project which does everything in the main thread and draws on a canvas: https://github.com/protomaps/protomaps.js

Both projects are basically using the idea of a Canvas for vector tile drawing.

I really hope OffscreenCanvas becomes mainstream one day!

**Comment by bchapuis** [comment]
Really excited by this initiative as well! @Drabble and I did similar experiments early this year and we are convinced that rust and wgpu will enable to develop truely portable web mapping solutions. Maplibre-rs is the perfect place to form a community around this topic.

![IMAGE 2022-04-23 13:54:14](https://user-images.githubusercontent.com/238747/164893426-2320cda1-3eb5-4269-a9f5-0ed47e9e683c.jpg)

**Reply by codeart1st** [reply]
I like what I see :+1:

**Reply by pka** [reply]
Same excitement here. In my [experiments](https://github.com/pka/flatgeobuf-bevy) I used the Bevy game engine instead of using WebGPU directly.

![flatgeobuf-wgpu](https://user-images.githubusercontent.com/19007/165043088-bb20c922-3dea-45a3-b740-edc9c51bb2ad.gif)

**Reply by codeart1st** [reply]
How big is the wasm filesize for this?

**Reply by pka** [reply]
I don't know, had to make it buildable again first...

**Reply by maxammann** [reply]
For maplibre-rs its right now 1.5M uncompressed which equals 800KB compressed.

**Reply by codeart1st** [reply]
It's relatively huge, but this is because of tracing and debug stuff, isn't it?

My is currently around 200-210kB uncompressed. Round about 70-75kB gzip. But less features

**Reply by maxammann** [reply]
That's very tiny. Its already a release build. At some point that needs to be investigated.
